mass-deployer/internal/parser/parser_test.go

```diff
@@ -0,0 +1,149 @@
+package parser
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+)
+
+func TestParseConfig(t *testing.T) {
+	t.Run("invaild yaml format", func(t *testing.T) {
+		conf := ` {
+  "node_groups": [
+    {
+      "nodes_count": 10
+    }
+} `
+		_, err := ParseConfig([]byte(conf))
+		assert.Error(t, err)
+	})
+
+	t.Run("invaild node_group", func(t *testing.T) {
+		conf := `
+vms:
+  - name: example-vm
+    vms_count: 2
+    node_group: example-group
+ssh_keys: 
+  example1: ssh-key1
+mnemonic: example-mnemonic
+network: dev
+    `
+		_, err := ParseConfig([]byte(conf))
+		assert.Error(t, err)
+	})
+
+	t.Run("invaild vms", func(t *testing.T) {
+		conf := `
+node_groups:
+  - name: example-group
+    nodes_count: 3
+    free_cpu: 8
+    free_mru: 16384
+ssh_keys: 
+  example1: ssh-key1
+mnemonic: example-mnemonic
+network: dev
+`
+		_, err := ParseConfig([]byte(conf))
+		assert.Error(t, err)
+	})
+
+	t.Run("invaild ssh_keys", func(t *testing.T) {
+		conf := `
+node_groups:
+  - name: example-group
+    nodes_count: 3
+    free_cpu: 8
+    free_mru: 16384
+vms:
+  - name: example-vm
+    vms_count: 2
+    node_group: example-group
+    cpu: 2
+    mem: 4096
+    flist: example-flist
+    root_size: 20
+    entry_point: /sbin/zinit init
+mnemonic: example-mnemonic
+network: dev
+    `
+		_, err := ParseConfig([]byte(conf))
+		assert.Error(t, err)
+	})
+
+	t.Run("invaild mnemonic", func(t *testing.T) {
+		conf := `
+node_groups:
+  - name: example-group
+    nodes_count: 3
+    free_cpu: 8
+    free_mru: 16384
+vms:
+  - name: example-vm
+    vms_count: 2
+    node_group: example-group
+    cpu: 2
+    mem: 4096
+    flist: example-flist
+    root_size: 20
+    entry_point: /sbin/zinit init
+ssh_keys: 
+  example1: ssh-key1
+network: dev
+    `
+		_, err := ParseConfig([]byte(conf))
+		assert.Error(t, err)
+	})
+
+	t.Run("invaild network", func(t *testing.T) {
+		conf := `
+node_groups:
+  - name: example-group
+    nodes_count: 3
+    free_cpu: 8
+    free_mru: 16384
+vms:
+  - name: example-vm
+    vms_count: 2
+    node_group: example-group
+    cpu: 2
+    mem: 4096
+    flist: example-flist
+    root_size: 20
+    entry_point: /sbin/zinit init
+ssh_keys: 
+  example1: ssh-key1
+network: example-network
+mnemonic: example-mnemonic
+    `
+		_, err := ParseConfig([]byte(conf))
+		assert.Error(t, err)
+	})
+
+	t.Run("valid config", func(t *testing.T) {
+		conf := `
+node_groups:
+  - name: example-group
+    nodes_count: 3
+    free_cpu: 8
+    free_mru: 16384
+vms:
+  - name: example-vm
+    vms_count: 2
+    node_group: example-group
+    cpu: 2
+    mem: 4096
+    flist: example-flist
+    root_size: 20
+    entry_point: /sbin/zinit init
+    ssh_key: example1
+ssh_keys: 
+  example1: ssh-key1
+network: dev
+mnemonic: rival oyster defense garbage fame disease mask mail family wire village vibrant index fuel dolphin
+    `
+		_, err := ParseConfig([]byte(conf))
+		assert.NoError(t, err)
+	})
+}
```

Comment:
Can be expanded with more descriptive test names
- single zone single vm
- single zone many vms
- no zones single vm
- empty configs
- zone with < vm required resources 
- wrong filters (wrong region, country, invalid farm, .. etc)
- special characters in names used (includugin spaces)
- check for root disk 
- flists cases (.flist, .fl) and .flist that doesn't work when you append .md5 to the url and empty one
- cases for sshkeys (in the integration tests we need to see if they get added correctly to the machines)
- cpu/ram limits

could be more

Reply:
do you mean by zone timezone here?
I added more tests for the limits and the flists
the other tests can be created in the deployer as the parser doesn't check if the vm resources actually exist